pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceTagPoolSelector.java

```diff
@@ -0,0 +1,135 @@
+    LOGGER.info("{} enabled instances have the tag: {} for table: {}", numCandidateInstances, tag, _tableNameWithType);
+
+    Map<Integer, List<InstanceConfig>> poolToInstanceConfigsMap = new TreeMap<>();
+    if (_tagPoolConfig.isPoolBased()) {
+      // Pool based selection
+
+      // Extract the pool information from the instance configs
+      for (InstanceConfig instanceConfig : candidateInstanceConfigs) {
+        Map<String, String> poolMap = instanceConfig.getRecord().getMapField(Instance.POOL_KEY);
+        if (poolMap != null && poolMap.containsKey(tag)) {
+          int pool = Integer.parseInt(poolMap.get(tag));
+          poolToInstanceConfigsMap.computeIfAbsent(pool, k -> new ArrayList<>()).add(instanceConfig);
+        }
+      }
+      Preconditions
+          .checkState(!poolToInstanceConfigsMap.isEmpty(), "No enabled instance has the pool configured for the tag: %s",
+              tag);
+      Map<Integer, Integer> poolToNumInstancesMap = new TreeMap<>();
+      for (Map.Entry<Integer, List<InstanceConfig>> entry : poolToInstanceConfigsMap.entrySet()) {
+        poolToNumInstancesMap.put(entry.getKey(), entry.getValue().size());
+      }
+      LOGGER.info("Number instances for each pool: {} for table: {}", poolToNumInstancesMap, _tableNameWithType);
+
+      // Calculate the pools to select based on the selection config
+      Set<Integer> pools = poolToInstanceConfigsMap.keySet();
+      List<Integer> poolsToSelect = _tagPoolConfig.getPools();
+      if (poolsToSelect != null && !poolsToSelect.isEmpty()) {
+        Preconditions.checkState(pools.containsAll(poolsToSelect), "Cannot find all instance pools configured: %s",
+            poolsToSelect);
+      } else {
+        int numPools = poolToInstanceConfigsMap.size();
+        int numPoolsToSelect = _tagPoolConfig.getNumPools();
+        if (numPoolsToSelect > 0) {
+          Preconditions
+              .checkState(numPoolsToSelect <= numPools, "Not enough instance pools (%s in the cluster, asked for %s)",
+                  numPools, numPoolsToSelect);
+        } else {
+          numPoolsToSelect = numPools;
+        }
+
+        // Directly return the map if all the pools are selected
+        if (numPools == numPoolsToSelect) {
+          LOGGER.info("Selecting all {} pools: {} for table: {}", numPools, pools, _tableNameWithType);
+          return poolToInstanceConfigsMap;
+        }
+
+        // Select pools based on the table name hash to evenly distribute the tables
+        poolsToSelect = new ArrayList<>(numPoolsToSelect);
+        List<Integer> poolsInCluster = new ArrayList<>(pools);
+        for (int i = 0; i < numPoolsToSelect; i++) {
+          poolsToSelect.add(poolsInCluster.get((_tableNameHash + i) % numPools));
+        }
+      }
+
+      // Keep the pools selected
+      LOGGER.info("Selecting pools: {} for table: {}", poolsToSelect, _tableNameWithType);
+      pools.retainAll(poolsToSelect);
```

Comment:
Why do we need to modify `pools` here? it seems that it's a local variable and it's not used after here?

Reply:
`pools` is the keySet() of the `poolToInstanceConfigsMap`, so any change to `pools` will reflect to the map.

Reply:
ah i see :)